ZEP000-wf-stream-processing.md

```diff
@@ -0,0 +1,1421 @@
+
+### Event Sub Process
+
+The event sub process element can be triggered by timer, message and error
+events. It can be either interrupting or non-interrupting. The non-interrupting
+form is relatively simple. Commands that activate event elements on processing
+(i.e. `Timer:Trigger` for timer events or
+`WorkflowInstanceSubscription:Correlate` for message events), simply activate
+the corresponding start event element of the event sub process by writing a
+`StartEvent:Activate_Element` command, thereby forking the execution of that
+workflow instance.
+
+Interrupting event sub processes require similar behavior as interrupting
+boundary event events, because they are logically equivalent to interrupting
+boundary events attached to the enclosing process.
+
+#### Interrupting Error Event Sub Process
+Command to Process | Written Events and Follow-up Commands
+---|---
+Job:Throw_Error | Job:Error_Thrown, ServiceTask:Terminate_Element (main)
+ServiceTask:Terminate_Element (main) | ServiceTask:Element_Terminating (main), ServiceTask:Element_Terminated (main), StartEvent:Activate_Element (sub)
+
+Notes:
+* an error can also be thrown by an error end event, that can be processed
+  through `EndEvent:Activate_Element` which results in
+  `EndEvent:Element_Activating`, `EndEvent:Element_Activated` and
+  `StartEvent:Activate_Element`
+
+On processing:
+* the `Job:Throw_Error` command marks the error on the job, makes the job non-activatable
+* the `ServiceTask:Terminate_Element` command transitions the service task element to terminated
+
+On replay:
+* the `Job:Error_Thrown` event marks the error on the job, makes the job non-activatable
+* the `ServiceTask:Element_Terminating` event transitions the service task element to terminating
+* the `ServiceTask:Element_Terminated` event transitions the service task element to terminated
+* processing can start again at `StartEvent:Activate_Element`
+
+Differences to current engine:
+* currently, the `Job:Error_Thrown` record is processed resulting in an event
+  trigger and the `Event_Occurred` record on the start event
+
+Considered alternatives:
+* nothing to mention
+
+### Message Start Event
+
+Command to Process | Written Events and Follow-up Commands
+---|---
+MessageStartEventSubscription:Open | MessageStartEventSubscription:Opened
+Message:Publish | Message:Published, MessageStartEventSubscription:Correlated, [Message:Deleted], Process:Activate_Element
+
+Notes:
+* the above table assumes the message is correlated to a message start event
+  subscription, if it isn't then it might be correlated to another message
+  event, or be ignored and deleted without correlation
+* `WorkflowInstanceSubscription` is not necessary like in message intermediate
+  catch event, because a new workflow instance can be created on any partition
+* the `Message:Deleted` event is only written when the message is not allowed to
+  live after it was published (i.e. `ttl = 0`). However, it will also be written
```

Comment:
When the `ttl > 0` then there is indeed an outside process that writes a `Message:Delete` command, that is then processed into `Message:Deleted`.

>Also I think the name Message:Deleted is misleading in this context

You're absolutely right. I'd prefer a `Message:Ignored` as wel for `ttl = 0` message that are not correlated. However, it changes the current behaviour. Currently, the `Message:Deleted` is written immediately when the `ttl = 0`. We should question whether we want to change that now.

Reply:
If it is an external process that does not use commands, then this might be dangerous during replay, E.g. if this independent process deletes timed out messages before the events that reference these messages have been replayed

Reply:
I agree that the name is not exact. But introducing a new intent especially for this case would increase the complexity of the lifecycle. I prefer to keep it simple. 

Renaming `Message:Deleted` to `Message:Expired` in all cases would work. I see it as an optional improvement.

Let's add it under open questions :writing_hand: